api/src/main/java/org/apache/iceberg/transforms/Timestamps.java

```diff
@@ -31,54 +32,117 @@
 import org.apache.iceberg.util.DateTimeUtil;
 import org.apache.iceberg.util.SerializableFunction;
 
-enum Timestamps implements Transform<Long, Integer> {
-  YEAR(ChronoUnit.YEARS, "year"),
-  MONTH(ChronoUnit.MONTHS, "month"),
-  DAY(ChronoUnit.DAYS, "day"),
-  HOUR(ChronoUnit.HOURS, "hour");
+class Timestamps implements Transform<Long, Integer> {
+
+  static final Timestamps YEAR_FROM_MICROS = new Timestamps(ChronoUnit.MICROS, ChronoUnit.YEARS);
+  static final Timestamps MONTH_FROM_MICROS = new Timestamps(ChronoUnit.MICROS, ChronoUnit.MONTHS);
+  static final Timestamps DAY_FROM_MICROS = new Timestamps(ChronoUnit.MICROS, ChronoUnit.DAYS);
+  static final Timestamps HOUR_FROM_MICROS = new Timestamps(ChronoUnit.MICROS, ChronoUnit.HOURS);
+  static final Timestamps YEAR_FROM_NANOS = new Timestamps(ChronoUnit.NANOS, ChronoUnit.YEARS);
+  static final Timestamps MONTH_FROM_NANOS = new Timestamps(ChronoUnit.NANOS, ChronoUnit.MONTHS);
+  static final Timestamps DAY_FROM_NANOS = new Timestamps(ChronoUnit.NANOS, ChronoUnit.DAYS);
+  static final Timestamps HOUR_FROM_NANOS = new Timestamps(ChronoUnit.NANOS, ChronoUnit.HOURS);
+
+  static Timestamps get(Types.TimestampType type, String resultTypeUnit) {
+    switch (resultTypeUnit.toLowerCase(Locale.ENGLISH)) {
+      case "year":
+        return get(type, ChronoUnit.YEARS);
+      case "month":
+        return get(type, ChronoUnit.MONTHS);
+      case "day":
+        return get(type, ChronoUnit.DAYS);
+      case "hour":
+        return get(type, ChronoUnit.HOURS);
+      default:
+        throw new IllegalArgumentException(
+            "Unsupported source/result type units: " + type + "->" + resultTypeUnit);
+    }
+  }
+
+  static Timestamps get(Types.TimestampType type, ChronoUnit resultTypeUnit) {
+    switch (type.typeId()) {
+      case TIMESTAMP:
+        switch (resultTypeUnit) {
+          case YEARS:
+            return YEAR_FROM_MICROS;
+          case MONTHS:
+            return MONTH_FROM_MICROS;
+          case DAYS:
+            return DAY_FROM_MICROS;
+          case HOURS:
+            return HOUR_FROM_MICROS;
+        }
+        break;
+      case TIMESTAMP_NANO:
```

Comment:
When would a `TimestampType` ever have `Type.TypeID.TIMESTAMP_NANO`? Wouldn't that only happen for `TimestampNanoType`?

Reply:
Fixed. I missed this when separating out the type.